### PR TITLE
Allows a Shadow to search for and add themself to an interview.

### PR DESCRIPTION
**Pull request by meagancheese**
Let me know if the text in the modals needs to be changed more to be Shadow Specific.

Screenshot 2020-08-04 at 3 05 46 PM
Screenshot 2020-08-04 at 3 20 13 PM
Screenshot 2020-08-04 at 3 20 24 PM
Screenshot 2020-08-04 at 3 20 33 PM
Screenshot 2020-08-04 at 3 20 42 PM
Screenshot 2020-08-04 at 3 20 54 PM

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 1]
```diff
@@ -169,6 +171,68 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
     }
   }
 
+  // Send the request's contents to Datastore in the form of an updated ScheduledInterview object.
+  // Adds a shadow.
```
Comment (apluscs) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Nit: specify it adds current user as the shadow

File: src/main/java/com/google/sps/data/DatastoreScheduledInterviewDao.java [hunk 2]
```diff
@@ -73,13 +73,37 @@ public DatastoreScheduledInterviewDao() {
     return scheduledInterviews;
   }
 
+  /**
+   * Returns a list, sorted by start time, of all ScheduledInterview objects ranging from minTime to
+   * maxTime that are for the selected position and do not already have a shadow.
+   */
+  public List<ScheduledInterview> getForPositionWithoutShadowInRange(
+      Job position, Instant minTime, Instant maxTime) {
+    List<Filter> allFilters = new ArrayList<>();
```
Comment (apluscs) [inline src/main/java/com/google/sps/data/DatastoreScheduledInterviewDao.java:82]: Kind of a nit: You can further minimize the logic in getPossibleInterviews() in the Load servlet by adding more filters and a shadowId param to this function. Use notEqual filters to perform this logic:
!(interviewerId == userId) && !(intervieweeId == userId) && !(shadow == userId)

File: src/main/webapp/profile.js [hunk 3]
```diff
@@ -34,14 +34,15 @@ function submitProfileForm(methodType, redirectUrl) {
...
+  okShadow = document.getElementById('ok-shadow-check').checked;
   const personJson = {
     firstName: $('#first-name-field').val(),
     lastName: $('#last-name-field').val(),
     company: $('#company-field').val(),
     job: $('#job-field').val(),
     linkedIn: $('#linkedin-field').val(),
     qualifiedJobs: qualifiedJobs,
-    okShadow: $('#ok-shadow-check').val()
+    okShadow: okShadow
```
Comment (apluscs) [inline src/main/webapp/profile.js]: Nit: just in line this like the rest of the fields
l prefer jQuery bc everything else is jQuery but if it's not worth it don't bother
Reference if you need it: https://stackoverflow.com/questions/901712/how-do-i-check-whether-a-checkbox-is-checked-in-jquery

File: src/main/webapp/search-interview.js [hunk 4]
```diff
@@ -21,16 +21,30 @@ function onSearchInterviewLoad() {
...
-  fetch(`/load-interviews?timeZoneOffset=${browserTimezoneOffset()}&position=${selectedPosition()}`)
-  .then(response => response.text())
-  .then(interviewTimes => {
-    interviewTimesDiv().innerHTML = interviewTimes;
-  });
+  const searchResultsDiv = document.getElementById('search-results');
+  searchResultsDiv.removeAttribute('hidden');
+  const role = selectedRole();
+  if (role === 'Interviewee') {
+    fetch(`/load-interviews?timeZoneOffset=${browserTimezoneOffset()}&position=${selectedEnumPosition()}`)
```
Comment (apluscs) [inline src/main/webapp/search-interview.js]: Nit: these two cases are basically the same. l'd say define a string that's either 'load-interviews' or 'shadow-load-interviews' and put that into the fetch URL

File: src/test/java/com/google/sps/data/DatastoreScheduledInterviewDaoTest.java [hunk 5]
```diff
@@ -190,6 +190,94 @@ public void getForPerson() {
     Assert.assertEquals(expected, result);
   }
 
+  // Tests that only interviews without a shadow are returned.
+  @Test
```
Comment (apluscs) [inline src/test/java/com/google/sps/data/DatastoreScheduledInterviewDaoTest.java:194]: lf you add the other three filters as specified in my first comment make sure to add tests here

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 6]
```diff
@@ -169,6 +171,68 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (apluscs) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:170]: Not confident this is being called, correct me if l'm wrong. l see when someone selects an interview `selectInterview(this)` is called, but that calls the doPost()
https://github.com/googleinterns/InterviewMe/blob/5ebc6552804860cede838e785f2170cc0e720061/src/main/webapp/search-interview.js
Or if you planned to do this in another PR that's good too

**Review by apluscs — CHANGES_REQUESTED** [review]
Small things

File: src/main/java/com/google/sps/data/DatastoreScheduledInterviewDao.java (hunk 2 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/data/DatastoreScheduledInterviewDao.java:82]: I tried it, it doesn't work because you're not allowed to have NOT_EQUAL filters on several parameters within the same Query. Definitely good thought though.

File: src/test/java/com/google/sps/data/DatastoreScheduledInterviewDaoTest.java (hunk 5 above)
Comment (meagancheese) [reply src/test/java/com/google/sps/data/DatastoreScheduledInterviewDaoTest.java:194]: Tried it and found that it was not possible (sadly).

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 1 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/main/webapp/profile.js (hunk 3 above)
Comment (meagancheese) [reply src/main/webapp/profile.js]: Done

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 6 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:170]: This is called within SelectInterview (the Shadow portion).

File: src/main/webapp/search-interview.js (hunk 4 above)
Comment (meagancheese) [reply src/main/webapp/search-interview.js]: Removed repeats.

**Comment by meagancheese** [comment]
Null name bug is back...

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 7]
```diff
@@ -226,8 +290,10 @@ private ScheduledInterviewRequest makeScheduledInterviewRequest(
             .get(scheduledInterview.intervieweeId())
             .map(Person::firstName)
             .orElse("Nonexistent User");
-    String shadow =
-        personDao.get(scheduledInterview.shadowId()).map(Person::firstName).orElse("None");
+    String shadow = "None";
+    if (!scheduledInterview.shadowId().equals("")) {
```
Comment (apluscs) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:283]: Some docu explaining why this check is here would be nice.

File: src/main/webapp/search-interview.js [hunk 8]
```diff
@@ -54,30 +66,60 @@ function selectInterview(interviewer) {
...
+      }
+      let requestObject = {
+        company: company,
+        job: job,
+        utcStartTime: utcStartTime,
+        position: selectedEnumPosition()
+      };
+      let requestBody = JSON.stringify(requestObject);
+      let request = new Request('/scheduled-interviews', {method: 'PUT', body: requestBody});
+      fetch(request).then(() => {window.location.replace('/scheduled-interviews.html');});
```
Comment (apluscs) [inline src/main/webapp/search-interview.js]: All this can be abstracted into an updateDatastore() method, the only change is the methodType which can be a param.

**Review by apluscs — APPROVED** [review]

**Review by apluscs — CHANGES_REQUESTED** [review]
wrong button

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 7 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:283]: Done

File: src/main/webapp/search-interview.js (hunk 8 above)
Comment (meagancheese) [reply src/main/webapp/search-interview.js]: Done

**Review by apluscs — APPROVED** [review]

**Review by GrantComm — APPROVED** [review]
LGTM

File: src/main/java/com/google/sps/data/FakeScheduledInterviewDao.java [hunk 9]
```diff
@@ -84,6 +85,23 @@ public FakeScheduledInterviewDao() {
...
+  /**
+   * Returns a list, sorted by start time, of all ScheduledInterview objects ranging from minTime to
+   * maxTime that are for the selected position and do not already have a shadow.
+   */
+  public List<ScheduledInterview> getForPositionWithoutShadowInRange(
+      Job position, Instant minTime, Instant maxTime) {
+    List<ScheduledInterview> interviewsInRange = getInRange(minTime, maxTime);
+    Set<ScheduledInterview> notValidInterviews = new HashSet<ScheduledInterview>();
+    for (ScheduledInterview interview : interviewsInRange) {
+      if (!interview.position().equals(position) || !interview.shadowId().equals("")) {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/data/FakeScheduledInterviewDao.java:97]: shadowId().equals("") is not equivalent to not having a shadow?

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 10]
```diff
@@ -226,8 +290,13 @@ private ScheduledInterviewRequest makeScheduledInterviewRequest(
...
             .map(Person::firstName)
             .orElse("Nonexistent User");
-    String shadow =
-        personDao.get(scheduledInterview.shadowId()).map(Person::firstName).orElse("None");
+    // When an interview is first scheduled, the shadowId is set to an empty string. Since this
+    // behaviour is expected, here we prevent a null or empty name exception with creating keys
+    // in datastore.
+    String shadow = "None";
+    if (!scheduledInterview.shadowId().equals("")) {
+      shadow = personDao.get(scheduledInterview.shadowId()).map(Person::firstName).orElse("None");
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Kind of clunky to duplicate the "None" thing.

Maybe just:
```java
    if (!scheduledInterview.shadowId().equals("")) {
      Optional<String> firstName = personDao.get(scheduledInterview.shadowId()).map(Person::firstName);
      if (firstName.hasValue()) {
        shadow = firstName.get();
      }
    }
```

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+  public void init(
+      ScheduledInterviewDao scheduledInterviewDao, PersonDao personDao, Instant currentTime) {
+    this.scheduledInterviewDao = scheduledInterviewDao;
+    this.personDao = personDao;
+    this.currentTime = currentTime;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java:79]: long line

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+    // The user will be shown available interview times for the next four weeks, starting from the
+    // current time.
+    TimeRange interviewSearchTimeRange =
+        new TimeRange(utcTime.toInstant(), utcTime.toInstant().plus(27, ChronoUnit.DAYS));
+    String position = request.getParameter("position");
+    Job selectedPosition = Job.valueOf(Job.class, position);
+    UserService userService = UserServiceFactory.getUserService();
+    String userEmail = userService.getCurrentUser().getEmail();
+    String userId = userService.getCurrentUser().getUserId();
+    // Since Users returned from the LocalUserService (in tests) do not have userIds, here we set
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java:98]: This hack, particularly the repeated copy pasting of it, is probably the worst thing we've allowed in this code.

(Just a long term lesson -- this project is _full_ of terrible stuff that we've permitted for the sake of expediency. Please don't reflect on this experience as setting the bar for code quality!!).

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+      possibleInterviewSlots.add(
+          PossibleInterviewSlot.create(
+              interview.when().start().toString(),
+              getDate(interview.when().start(), timezoneOffset),
+              getTime(interview.when().start(), timezoneOffset)));
+    }
+    return possibleInterviewSlots;
+  }
+
+  private String getDate(Instant instant, ZoneOffset timezoneOffset) {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: I feel like I've seen this function (and the next two functions) many many times. Is this duplicated from somewhere?

File: src/main/webapp/search-interview.js [hunk 14]
```diff
@@ -54,48 +66,76 @@ function selectInterview(interviewer) {
...
-    };
-    let requestBody = JSON.stringify(requestObject);
-    let request = new Request('/scheduled-interviews', {method: 'POST', body: requestBody});
-    fetch(request).then(() => {window.location.replace('/scheduled-interviews.html');});
   }
 }
 
+function updateDatastore(method, company, job, utcStartTime) {
+  if (company === '<Not specified>') {
+    company = '';
```
Comment (acmcarther) [inline src/main/webapp/search-interview.js:98]: Having multiple representations of "not specified" is a pretty bad code smell, but I know we're not going to have time to fix this.

**Review by acmcarther — CHANGES_REQUESTED** [review]
Did this PR just copy past the interviewer servlet whole sale? That's pretty shameless.

File: src/main/java/com/google/sps/data/FakeScheduledInterviewDao.java (hunk 9 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/data/FakeScheduledInterviewDao.java:97]: It is equivalent

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 11 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java:79]: Formatter inlines it when I try and change it.

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 12 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java:98]: Yeah, we probably should have come up with a better way to set current user in tests earlier on. Basically making our own mock of user service.

File: src/main/webapp/search-interview.js (hunk 14 above)
Comment (meagancheese) [reply src/main/webapp/search-interview.js:98]: If it were to be fixed, we'd put <Not specified> everywhere instead of an empty string and then we'd have to deal with the HTML processing it as a tag again.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 10 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 13 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: I removed the duplication that I saw as obvious, let me know if there is more that I should address.

File: src/main/java/com/google/sps/data/FakeScheduledInterviewDao.java (hunk 9 above)
Comment (acmcarther) [reply src/main/java/com/google/sps/data/FakeScheduledInterviewDao.java:97]: Ah I see, misread the code I guess.

File: src/main/java/com/google/sps/data/AvailabilityTimeSlotGenerator.java [hunk 15]
```diff
@@ -196,12 +196,12 @@ private static Instant getStartOrEndOfWeek(ZonedDateTime day, boolean start) {
   }
 
   // Uses an Instant and a timezoneOffsetMinutes int to create a ZonedDateTime instance.
-  private static ZonedDateTime generateDay(Instant instant, int timezoneOffsetMinutes) {
+  public static ZonedDateTime generateDay(Instant instant, int timezoneOffsetMinutes) {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/data/AvailabilityTimeSlotGenerator.java]: These helper functions should not live in AvailabilityTimeSlotGenerator anymore now that it is public... it doesn't make sense at call sites to be invoking AvailabilityTimeSlotGenerator.generateDay. Move this into a "TimeUtils" or something like that.

File: src/main/java/com/google/sps/servlets/LoadInterviewsServlet.java [hunk 16]
```diff
@@ -218,21 +210,26 @@ private boolean isAnHourWorthOfSlots(
         .equals(availabilities.get(index + numberOfSlotsAfterFirstInAnHour).when().start());
   }
 
-  private String getDate(Instant instant, ZoneOffset timezoneOffset) {
+  // Uses an Instant and a timezoneOffset to create a ZonedDateTime instance.
+  static ZonedDateTime generateDay(Instant instant, ZoneOffset timezoneOffset) {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/LoadInterviewsServlet.java]: This function and others like it should probably live in the "TimeUtils" as well.

File: src/main/java/com/google/sps/servlets/LogInServlet.java [hunk 17]
```diff
@@ -35,21 +26,20 @@ public static LoginInfo getLoginInfo(String rootUrl) {
     boolean loggedIn = userService.isUserLoggedIn();
     LoginInfo toSend;
     if (loggedIn) {
-      return new LoginInfo(
-          loggedIn, userService.createLogoutURL(rootUrl), userService.getCurrentUser().getEmail());
+      return new LoginInfo(loggedIn, "", userService.getCurrentUser().getEmail());
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/LogInServlet.java]: These changes look random... I'm guessing you synced or something though.

File: src/main/java/com/google/sps/servlets/LogoutServlet.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import javax.servlet.http.Cookie;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.Set;
+
+/** Servlet that handles logout. */
+@WebServlet("/logout")
+public class LogoutServlet extends HttpServlet {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/LogoutServlet.java]: Why is my diff showing this file?

(It's not easy to review a large cl like this one when random upstream changes fall into it.)

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 19]
```diff
@@ -169,6 +172,68 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+  // Adds the current user as a shadow.
+  @Override
+  public void doPut(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String shadowEmail = userService.getCurrentUser().getEmail();
+    String shadowId = getUserId();
+
+    InterviewPostOrPutRequest putRequest;
+    try {
+      putRequest = new Gson().fromJson(getJsonString(request), InterviewPostOrPutRequest.class);
+    } catch (Exception JsonSyntaxException) {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Don't catch arbitrary `Exception`. This will catch things like "Out of memory" which should really just crash the application.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 20]
```diff
@@ -169,6 +172,68 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
...
+      response.sendError(400);
+      return;
+    }
+    if (!putRequest.allFieldsPopulated()) {
+      response.sendError(400);
+      return;
+    }
+
+    String interviewerCompany = putRequest.getCompany();
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Move these variables closer to the usage site.

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java [hunk 21]
```diff
@@ -226,8 +291,17 @@ private ScheduledInterviewRequest makeScheduledInterviewRequest(
             .get(scheduledInterview.intervieweeId())
             .map(Person::firstName)
             .orElse("Nonexistent User");
-    String shadow =
-        personDao.get(scheduledInterview.shadowId()).map(Person::firstName).orElse("None");
+    // When an interview is first scheduled, the shadowId is set to an empty string. Since this
+    // behaviour is expected, here we prevent a null or empty name exception with creating keys
+    // in datastore.
+    String shadow = "None";
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:282]: Use `""` instead?

Would prefer to avoid having multiple notations for "no shadow".

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 11 above)
Comment (acmcarther) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java:79]: Oh well I guess.

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 12 above)
Comment (acmcarther) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java:98]: My "long-term" preference I think would be for us to not use `userId` as an internal id. 

I consulted the documentation on this recently and basically this quasi-internal. I suspect the bad behavior of the test userservice (in not setting it) is in part to dissuade users from doing what we did in making this id significant in the application.

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+    // Since Users returned from the LocalUserService (in tests) do not have userIds, here we set
+    // the userId equal to a hashcode.
+    if (userId == null) {
+      userId = String.format("%d", userEmail.hashCode());
+    }
+
+    List<ScheduledInterview> possibleInterviews =
+        getPossibleInterviews(
+            scheduledInterviewDao, selectedPosition, interviewSearchTimeRange, personDao, userId);
+
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: What semantic meaning do these blank newlines carry?

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+    int timezoneOffsetMinutes = Integer.parseInt(request.getParameter("timeZoneOffset"));
+    Preconditions.checkArgument(
+        Math.abs(timezoneOffsetMinutes) <= maxTimezoneOffsetMinutes,
+        "Offset greater than %d minutes (%d hours): %d",
+        maxTimezoneOffsetMinutes,
+        maxTimezoneOffsetHours,
+        timezoneOffsetMinutes);
+    ZoneOffset timezoneOffset =
+        AvailabilityTimeSlotGenerator.convertIntToOffset(timezoneOffsetMinutes);
+    ZonedDateTime day = LoadInterviewsServlet.generateDay(currentTime, timezoneOffset);
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: This variable isn't really doing much work. Inline into next line?

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+        scheduledInterviewsToPossibleInterviewSlots(possibleInterviews, timezoneOffset);
+
+    String date = possibleInterviewSlots.isEmpty() ? "" : possibleInterviewSlots.get(0).date();
+    List<ArrayList<PossibleInterviewSlot>> possibleInterviewsForWeek =
+        new ArrayList<ArrayList<PossibleInterviewSlot>>();
+
+    if (!possibleInterviewSlots.isEmpty()) {
+      ArrayList<PossibleInterviewSlot> dayOfSlots = new ArrayList<PossibleInterviewSlot>();
+      for (PossibleInterviewSlot possibleInterview : possibleInterviewSlots) {
+        if (!possibleInterview.date().equals(date)) {
```
Comment (acmcarther) [inline src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: This code could benefit from comments (particularly the resetting of `dayOfSlots`).

File: src/main/webapp/availability.js [hunk 25]
```diff
@@ -59,7 +59,11 @@ function updateAvailability() {
   };
   let requestBody = JSON.stringify(requestObject);
   let request = new Request('/availability', {method: 'PUT', body: requestBody});
-  fetch(request).then(() => location.reload()).catch((error) => {
+  fetch(request).then(() => {
+      if (reload) {
```
Comment (acmcarther) [inline src/main/webapp/availability.js:78]: Indentation is weird here. 

Move the `.then` down?

File: src/main/webapp/availability.js [hunk 26]
```diff
@@ -73,6 +77,7 @@ function goBack() {
     page = 0;
     return;
   }
+  updateAvailability(false);
```
Comment (acmcarther) [inline src/main/webapp/availability.js]: `updateAvailability(/*reload=*/false)`

(same below)

File: src/main/webapp/availability.js [hunk 27]
```diff
@@ -59,7 +59,11 @@ function updateAvailability() {
   };
   let requestBody = JSON.stringify(requestObject);
   let request = new Request('/availability', {method: 'PUT', body: requestBody});
-  fetch(request).then(() => location.reload()).catch((error) => {
+  fetch(request).then(() => {
+      if (reload) {
+        location.reload();
```
Comment (acmcarther) [inline src/main/webapp/availability.js:81]: This is really weird. Why are we doing this?

(i'm not likely to push back, but this probably at least merits a comment.)

File: src/main/webapp/search-interview.js [hunk 28]
```diff
@@ -14,23 +14,35 @@
...
-  searchResultsDiv.removeAttribute("hidden");
-  fetch(`/load-interviews?timeZoneOffset=${browserTimezoneOffset()}&position=${selectedPosition()}`)
-  .then(response => response.text())
-  .then(interviewTimes => {
-    interviewTimesDiv().innerHTML = interviewTimes;
-  });
+  const searchResultsDiv = document.getElementById('search-results');
+  searchResultsDiv.removeAttribute('hidden');
+  const role = selectedRole();
+  let servlet = '';
```
Comment (acmcarther) [inline src/main/webapp/search-interview.js:27]: Just FYI, this is one of those design decisions that would hurt a lot long term.

Basically, you're baking in an assumption that these two servlets accept the same request json and URL parameters. Without EndToEnd tests, this type of thing would break and only be seen in prod.

File: src/main/webapp/search-interview.js [hunk 29]
```diff
@@ -54,48 +66,76 @@ function selectInterview(interviewer) {
...
+        `You have scheduled a ${position} interview on ${date}` +
+        ` from ${time} with a ${company} ` +
+        `${job}. Check your email for more ` +
+        `information.`);
+      updateDatastore('POST', company, job, utcStartTime);
     }
-    if (job === '<Not specified>') {
-      job = '';
+  }
+  if (role === 'Shadow') {
```
Comment (acmcarther) [inline src/main/webapp/search-interview.js]: probably should be `else if`, not a big deal though.

**Review by acmcarther — APPROVED** [review]
Mostly nits

File: src/main/java/com/google/sps/servlets/LogoutServlet.java (hunk 18 above)
Comment (acmcarther) [reply src/main/java/com/google/sps/servlets/LogoutServlet.java]: Stale comment, resolving.

File: src/main/java/com/google/sps/servlets/LogInServlet.java (hunk 17 above)
Comment (acmcarther) [reply src/main/java/com/google/sps/servlets/LogInServlet.java]: Stale comment, resolving.

Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/LogInServlet.java]: Yep, not one of mine.

File: src/main/java/com/google/sps/servlets/LogoutServlet.java (hunk 18 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/LogoutServlet.java]: Handled

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 21 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java:282]: This is what is being sent to the page to be displayed under Shadow: on the Scheduled Interview Page. We decided "Shadow: None" looked better than "Shadow:"

File: src/main/java/com/google/sps/data/AvailabilityTimeSlotGenerator.java (hunk 15 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/data/AvailabilityTimeSlotGenerator.java]: Done

File: src/main/java/com/google/sps/servlets/LoadInterviewsServlet.java (hunk 16 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/LoadInterviewsServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 19 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java (hunk 20 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ScheduledInterviewServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 22 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: Removed

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 23 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: Done

File: src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java (hunk 24 above)
Comment (meagancheese) [reply src/main/java/com/google/sps/servlets/ShadowLoadInterviewsServlet.java]: Done

File: src/main/webapp/availability.js (hunk 25 above)
Comment (meagancheese) [reply src/main/webapp/availability.js:78]: Done

File: src/main/webapp/availability.js (hunk 26 above)
Comment (meagancheese) [reply src/main/webapp/availability.js]: Done

File: src/main/webapp/availability.js (hunk 27 above)
Comment (meagancheese) [reply src/main/webapp/availability.js:81]: Why are we reloading? To show the user that their update went through when they clicked the update button.

Comment (meagancheese) [reply src/main/webapp/availability.js:81]: Added a comment.

File: src/main/webapp/search-interview.js (hunk 28 above)
Comment (meagancheese) [reply src/main/webapp/search-interview.js:27]: I agree it would have to be decoupled as soon as they diverged. Claire asked for this here because there was so much repetition in the way we currently have it set up.

File: src/main/webapp/search-interview.js (hunk 29 above)
Comment (meagancheese) [reply src/main/webapp/search-interview.js]: Changed
